include/xtensor/xgenerator.hpp

```diff
@@ -0,0 +1,494 @@
+/***************************************************************************
+* Copyright (c) 2016, Johan Mabille and Sylvain Corlay                     *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+#ifndef xgenerator_HPP
+#define xgenerator_HPP
+
+#include <cstddef>
+#include <type_traits>
+#include <utility>
+#include <tuple>
+#include <algorithm>
+
+#include "xexpression.hpp"
+#include "xiterator.hpp"
+#include "xutils.hpp"
+
+namespace xt
+{
+
+    template <class F, class R>
+    class xgenerator_stepper;
+
+    /**************
+     * xgenerator *
+     **************/
+
+    /**
+     * @class xgenerator
+     * @brief Multidimensional function operating on indices.
+     *
+     * Th xgenerator class implements a multidimensional function
+     * operating on the supplied indices.
+     *
+     * @tparam F the function type
+     * @tparam R the return type of the function
+     */
+    template <class F, class R>
+    class xgenerator : public xexpression<xgenerator<F, R>>
+    {
+
+    public:
+
+        using self_type = xgenerator<F, R>;
+        using functor_type = typename std::remove_reference<F>::type;
+
+        using value_type = R;
+        using reference = value_type;
+        using const_reference = value_type;
+        using pointer = value_type*;
+        using const_pointer = const value_type*;
+        using size_type = std::size_t;
+        using difference_type = std::ptrdiff_t;
+
+        using shape_type = std::vector<size_type>;
+        using strides_type = std::vector<size_type>;
+        using closure_type = const self_type;
+
+        using const_stepper = xgenerator_stepper<F, R>;
+        using const_iterator = xiterator<const_stepper, shape_type>;
+        using const_storage_iterator = const_iterator;
+
+        template <class Func, class S>
+        xgenerator(Func&& f, const S& shape) noexcept;
+
+        size_type dimension() const noexcept;
+        shape_type shape() const;
```

Comment:
`shape()` should rather return a constant reference to `m_shape` which should solve potential dangling references issues.